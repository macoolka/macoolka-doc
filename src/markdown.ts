/**
 * markdown utilities
 * @file
 */

import * as O from 'fp-ts/lib/Option';
import { Module } from 'macoolka-type-model';
import { pipe } from 'fp-ts/lib/pipeable';
import { MonadDocument, printArticle, printJekyllPage } from 'macoolka-document';
import { fold, monoidString } from 'fp-ts/lib/Monoid';
import { MonidI18N, MonidI18NParam } from 'macoolka-i18n';
import * as A from 'fp-ts/lib/Array';
import { monadFormat } from 'macoolka-prettier';
import formatI18N from './i18n';
import { notEmptyArray } from 'macoolka-predicate';
import { MonadFunction } from 'macoolka-app/lib/MonadFunction';
import { Reader } from 'fp-ts/lib/Reader';
import * as path from 'path';
export const toc = require('markdown-toc');
export type I18nDocumentApp = Reader<MonadDocument, MonidI18N>;
export type MonadDocumentFunction = MonadFunction<MonadDocument, MonidI18N>;
const foldString = fold(monoidString);
function printElement(
  element:
    Module.MDocumentable & Module.MNameable & { _kind: Module.MIdentifierKind }): I18nDocumentApp {
  return C => i18n => {
    const name = element.deprecated ? C.doc.del(element.name) : element.name;
    const type = formatI18N({ id: `macoolka.data-model.${element._kind}` } as any)(i18n);

    const desc = pipe(
      element.description.map(C.doc.br),
      fold(monoidString),
      content => O.some({
        content: content,
      })
    );
    const signature = pipe(
      element.signature,
      O.fromNullable,
      O.map(value => ({
        title: formatI18N({ id: 'macoolka.doc.signature' })(i18n),
        content: C.doc.code('ts')(value),
      })
      )
    );
    const example = pipe(
      element.examples,
      O.fromPredicate(notEmptyArray),
      O.map(as => C.doc.code('ts')(as.join('\n'))),
      O.map(value => ({
        title: formatI18N({ id: 'macoolka.doc.example' })(i18n),
        content: value,
      }))

    );
    const since = pipe(
      element.since,
      O.fromNullable,
      O.map(value => ({
        content: formatI18N({ id: 'macoolka.doc.since', value: { value } })(i18n),
      })
      )
    );
    return printArticle({
      title: `${name} (${type})`,
      section: pipe(
        [
          desc,
          signature,
          example,
          since,
        ],
        A.compact
      ),
    })(C);

  };
}

function printClass(c: Module.MClass): I18nDocumentApp {
  return C => i18n => {
    return pipe(
      [
        printElement(c)(C)(i18n),
        pipe(
          [
            c.staticMethods.map(a => printElement(a)(C)(i18n)),
            c.methods.map(a => printElement(a)(C)(i18n)),
          ],
          A.flatten,
          foldString
        ),
      ],
      foldString
    );
  };

}

export function printToc(md: string): I18nDocumentApp {
  return C => i18n =>
    pipe(
      [
        C.doc.hr(),
        C.doc.section(`<h2 class="text-delta">${
          formatI18N({ id: 'macoolka.doc.contents' })(i18n)
          }</h2>`),
        C.doc.section(toc(md).content),
        C.doc.hr(),
      ],
      foldString
    );
}
interface PrintModuleInput extends MonadDocument {
  module: Module.MModule;
  counter: number;
  i18n: MonidI18NParam;
}
/**
 * @since 0.2.0
 */
export function printModule(C: PrintModuleInput) {

  const { module, counter, i18n } = C;

  const name = (module.path).join(path.sep);
  const modules = formatI18N({ id: 'macoolka.doc.modules' })(i18n);

  const header = printJekyllPage({ name, nav: { order: counter, parent: modules } })(C);
  const title = formatI18N({ id: 'macoolka.doc.overview' })(i18n);
  const description = pipe(
    module.description.map(C.doc.br),
    foldString
  );

  const moduleContent = printArticle({
    title,
    description,
  })(C);

  const body = pipe([
    module.interfaces.map(printElement),
    module.typealiases.map(printElement),
    module.classes.map(printClass),
    module.constants.map(printElement),
    module.functions.map(printElement),
    module.exports.map(printElement),
  ],
    A.flatten,
    as => as.map(a => a(C)(i18n)).join('')

  );

  const tocContent = printToc(body)(C)(i18n);

  return pipe(
    [
      header,
      moduleContent,
      tocContent,
      body,
    ],
    foldString,
    content => monadFormat({ parser: 'markdown', content })
  );

}
