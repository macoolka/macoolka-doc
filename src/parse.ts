/**
 * @file
 */
import { parse as _parse } from 'macoolka-type-ast';
import { Module } from 'macoolka-type-model';
import * as array from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/pipeable';
import * as E from 'fp-ts/lib/Either';
import * as R from 'fp-ts/lib/Reader';
import { FileWhereUniqueInput } from 'macoolka-store-core';
const filterModule = (a: Module.MModule) => {

    const result: Module.MModule = {
        ...a,
        constants: pipe(
            a.constants,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        functions: pipe(
            a.functions,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        classes: pipe(
            a.classes,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        interfaces: pipe(
            a.interfaces,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        typealiases: pipe(
            a.typealiases,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        exports: pipe(
            a.exports,
            array.filter(a => a.ignore !== true)
        ),
    };
    return result;
};
/**
 * Parse files to Module
 * @desczh
 * 解析文件到Module
 * @since 0.2.0
 */
export const parse = (files: FileWhereUniqueInput[]) =>
    pipe(
        _parse(files),
        E.map(as =>
            pipe(
                as,
                R.map(a =>
                    pipe(
                        a,
                        array.filter(a => a.file === true),
                        array.map(filterModule)
                    ))
            ))
    );
