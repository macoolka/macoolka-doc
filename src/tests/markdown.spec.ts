import { parse } from '../parse'
import { writeFileSync } from 'fs'
import { printModule } from '../markdown'
import { markdown } from 'macoolka-markdown'
import { pipe } from 'fp-ts/lib/pipeable'
import { array as A } from 'macoolka-collection'
import * as E from 'fp-ts/lib/Either'
import * as MN from 'macoolka-app/lib/MonadNode'
//import {formatWithType} from 'macoolka-format'
describe('markdowns', () => {
    it('print default', () => {

       const value= pipe(
            parse([{ container: __dirname, folders: ['fixtures'], name: 'common.ts' }]),
            E.map(as =>
                pipe(
                    as({}),
                    A.map(a =>
                        pipe(
                            printModule({ module: a, counter: 0, doc: markdown, i18n: {} }),
                            MN.map(result => {
                                writeFileSync(__dirname + '/common.md', result, 'utf-8')
                                expect(result).toMatchSnapshot();
                            })
                        )),
                    MN.sequence,
                    as=>as()
                ),
            ),
           
        )
        expect(E.isRight(value)).toBeTruthy()

    })
     it('print locale zh', () => {
        const value= pipe(
            parse([{ container: __dirname, folders: ['fixtures'], name: 'common.ts' }]),
            E.map(as =>
                pipe(
                    as({ i18n: { locale: 'zh' } }),
                    A.map(a =>
                        pipe(
                            printModule({ module: a, counter: 0, doc: markdown, i18n: { i18n: { locale: 'zh' } } }),
                            MN.map(result => {
                                writeFileSync(__dirname + '/common.md', result, 'utf-8')
                                expect(result).toMatchSnapshot();
                            })
                        )),
                    MN.sequence,
                    as=>as()
                ),
            ),
           
        )
        expect(E.isRight(value)).toBeTruthy()
     }) 

})