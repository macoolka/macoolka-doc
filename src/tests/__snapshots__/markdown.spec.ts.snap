// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdowns print default 1`] = `
"---
title: /home/fastspeeed/macoolka/macoolka-doc/src/tests/fixtures/common.ts
nav_order: 0
parent: Modules
---

# Overview

The define a document interface

It will be instanced by markdown or html

---

<h2 class=\\"text-delta\\">Table of contents</h2>

- [Document (interface)](#document-interface)
- [FormattingElement (interface)](#formattingelement-interface)
- [Header (interface)](#header-interface)
- [List (interface)](#list-interface)
- [markdown (constant)](#markdown-constant)
- [toc (constant)](#toc-constant)

---

# Document (interface)

The define a Document

**Signature**

\`\`\`ts
interface Document extends Header, List, FormattingElement {
  /**
   *Defines an article
   */
  article: (code: string) => string
  /**
   *Defines a section that is quoted from another source
   */
  blockquote: (code: string[]) => string
  /**
   *To create a line break (<br>).
   */
  br: () => string
  /**
   *Defines a piece of computer code
   */
  code: (language: string) => (a: string) => string
  /**
   *To create a horizontal rule,
   */
  hr: () => string
  /**
   *Defines an image
   */
  img: (text: string, href: string, title?: string) => string
  /**
   *To create a link, Defines the relationship between a document and an external resource
   */
  link: (text: string, href: string, title?: string) => string
  /**
   *To create paragraphs, use a blank line to separate one or more lines of text
   */
  paragraph: (code: string) => string
  /**
   *Defines a section in a document
   */
  section: (code: string) => string
  /**
   *Defines a table
   */
  table: (as: string[][]) => string
}
\`\`\`

Added in v0.2.0

# FormattingElement (interface)

Format a Element
You can add emphasis by making text bold or italic

**Signature**

\`\`\`ts
interface FormattingElement {
  /**
   *Defines important text
   */
  bold: (code: string) => string
  /**
   *The tag defines the title of a work (e.g. a book, a song, a movie, a TV show, a painting, a sculpture, etc.).
   *定义一个事物的标题
   */
  cite: (code: string) => string
  /**
   *Defines text that has been deleted from a document
   */
  del: (text: string) => string
  /**
   *a definition term
   */
  dfn: (code: string) => string
  /**
   *Defines emphasized text
   */
  em: (code: string) => string
  /**
   *The tag defines a text that has been inserted into a document.
   */
  ins: (code: string) => string
  /**
   *Defines italicize text.
   */
  italic: (code: string) => string
  /**
   *Defines marked text.
   *Use the tag if you want to highlight parts of your text.
   */
  mark: (code: string) => string
  /**
   *The tag defines a short quotation.
   */
  q: (code: string) => string
}
\`\`\`

**Example**

\`\`\`ts
const a = '1'
const b = '1'
const c = a === b
\`\`\`

Added in v0.2.0

# Header (interface)

create a heading
The number of number signs you use should correspond to the heading level

**Signature**

\`\`\`ts
interface Header {
  h1: (title: string) => string
  h2: (title: string) => string
  h3: (title: string) => string
  h4: (title: string) => string
  h5: (title: string) => string
  h6: (title: string) => string
}
\`\`\`

Added in v0.2.0

# List (interface)

The define a List

**Signature**

\`\`\`ts
interface List {
  /**
   *Defines an ordered list
   */
  ol: (as: string[]) => string
  /**
   *Defines an unordered list
   */
  ul: (as: string[]) => string
}
\`\`\`

Added in v0.2.0

# markdown (constant)

markdown instance on github

**Signature**

\`\`\`ts

export const markdown: Document = ...

\`\`\`

Added in v0.2.0

# toc (constant)

**Signature**

\`\`\`ts

export const toc: any = ...

\`\`\`

Added in v0.2.0
"
`;

exports[`markdowns print locale zh 1`] = `
"---
title: /home/fastspeeed/macoolka/macoolka-doc/src/tests/fixtures/common.ts
nav_order: 0
parent: 模块
---

# 概述

这定义了文章的接口。

将被 html、markdown 实现

---

<h2 class=\\"text-delta\\">目录</h2>

- [Document (接口)](#document-%E6%8E%A5%E5%8F%A3)
- [FormattingElement (接口)](#formattingelement-%E6%8E%A5%E5%8F%A3)
- [Header (接口)](#header-%E6%8E%A5%E5%8F%A3)
- [List (接口)](#list-%E6%8E%A5%E5%8F%A3)
- [markdown (常量)](#markdown-%E5%B8%B8%E9%87%8F)
- [toc (常量)](#toc-%E5%B8%B8%E9%87%8F)

---

# Document (接口)

定义一份文档

**签名**

\`\`\`ts
interface Document extends Header, List, FormattingElement {
  /**
   *标记内容为一篇文章
   */
  article: (code: string) => string
  /**
   *标记内容为引用别人的论述
   */
  blockquote: (code: string[]) => string
  /**
   *换行
   */
  br: () => string
  /**
   *标记内容为一段程序代码
   */
  code: (language: string) => (a: string) => string
  /**
   *水平线
   */
  hr: () => string
  /**
   *标记内容为图片
   */
  img: (text: string, href: string, title?: string) => string
  /**
   *标记内容为链接,一般指向外部资源
   */
  link: (text: string, href: string, title?: string) => string
  /**
   *标记内容为一个段落
   */
  paragraph: (code: string) => string
  /**
   *标记内容为一个章节
   */
  section: (code: string) => string
  /**
   *定义一个表格
   */
  table: (as: string[][]) => string
}
\`\`\`

v0.2.0 中添加

# FormattingElement (接口)

格式化一个元素，给一个元素加上特殊效果。

**签名**

\`\`\`ts
interface FormattingElement {
  /**
   *定义主要的文本
   */
  bold: (code: string) => string
  /**
   *The tag defines the title of a work (e.g. a book, a song, a movie, a TV show, a painting, a sculpture, etc.).
   *定义一个事物的标题
   */
  cite: (code: string) => string
  /**
   *定义一个删除标记的文本
   */
  del: (text: string) => string
  /**
   *定义一个术语
   */
  dfn: (code: string) => string
  /**
   *定义一个强调的文本
   */
  em: (code: string) => string
  /**
   *定义一个插入的文本
   */
  ins: (code: string) => string
  /**
   *定义一个斜体文本
   */
  italic: (code: string) => string
  /**
   *定义一个高亮的文本
   */
  mark: (code: string) => string
  /**
   *给文本加引号
   */
  q: (code: string) => string
}
\`\`\`

**示例**

\`\`\`ts
const a = '1'
const b = '1'
const c = a === b
\`\`\`

v0.2.0 中添加

# Header (接口)

定义了标题

数字为标题的级别

**签名**

\`\`\`ts
interface Header {
  h1: (title: string) => string
  h2: (title: string) => string
  h3: (title: string) => string
  h4: (title: string) => string
  h5: (title: string) => string
  h6: (title: string) => string
}
\`\`\`

v0.2.0 中添加

# List (接口)

列表定义

**签名**

\`\`\`ts
interface List {
  /**
   *有顺序列表
   */
  ol: (as: string[]) => string
  /**
   *无顺序列表
   */
  ul: (as: string[]) => string
}
\`\`\`

v0.2.0 中添加

# markdown (常量)

基于 github markdown

**签名**

\`\`\`ts

export const markdown: Document = ...

\`\`\`

v0.2.0 中添加

# toc (常量)

**签名**

\`\`\`ts

export const toc: any = ...

\`\`\`

v0.2.0 中添加
"
`;
